co-circom/co-circom-types/src/lib.rs

```
@@ -17,23 +17,77 @@ use serde::{Deserialize, Serialize};
+pub type Rep3SharedInput<F> = BTreeMap<String, Rep3InputType<F>>;
 
-/// A shorthand type for batched witnesses. Produced by the batched
-/// witness extension
-pub type BatchedWitness<P, S> = SharedWitness<Vec<P>, Vec<S>>;
+/// A batched REP3 shared input type. Should be used with the batched witness extension
+pub type BatchedRep3SharedInput<F> = BTreeMap<String, Vec<Rep3InputType<F>>>;
 
-/// A REP3 shared input type
-pub type Rep3SharedInput<F> = SharedInput<F, Rep3PrimeFieldShare<F>>;
+/// A shorthand type for batched witnesses. Produced by the batched witness extension
+pub type BatchedWitness<P, S> = SharedWitness<Vec<P>, Vec<S>>;
 
 /// A REP3 shared witness type
 pub type Rep3SharedWitness<F> = SharedWitness<F, Rep3PrimeFieldShare<F>>;
 
 /// A shamir shared witness type
 pub type ShamirSharedWitness<F> = SharedWitness<F, ShamirPrimeFieldShare<F>>;
 
+/// A REP3 input type
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
+pub enum Rep3InputType<F: PrimeField> {
+    /// The public variant
+    Public(
+        #[serde(
+            serialize_with = "mpc_core::serde_compat::ark_se",
+            deserialize_with = "mpc_core::serde_compat::ark_de"
+        )]
+        F,
+    ),
+    /// The shared variant
+    Shared(
+        #[serde(
+            serialize_with = "mpc_core::serde_compat::ark_se",
+            deserialize_with = "mpc_core::serde_compat::ark_de"
+        )]
+        Rep3PrimeFieldShare<F>,
+    ),
+}
+
+impl<F: PrimeField> Rep3InputType<F> {
+    /// Returns true if the input is public
+    pub fn is_public(&self) -> bool {
+        matches!(self, Rep3InputType::Public(_))
+    }
+
+    /// Returns the public input or None
+    pub fn into_public(self) -> Option<F> {
+        match self {
+            Rep3InputType::Public(value) => Some(value),
+            Rep3InputType::Shared(_) => None,
+        }
+    }
+
+    /// Returns the shared input or None
+    pub fn into_shared(self) -> Option<Rep3PrimeFieldShare<F>> {
```

This should return an error imo and not an option